### ISSUE TITLE
Loop

### ISSUE THREAD

**Issue by ibmua** [ducha-aiki/LSUVinit#6]
What's the reason for the loop? Does it make any mathematical sense? Is it looking for some sweetspot inside data, or what? You're essentially rescaling same weights on and on for neuron activations by standard deviation to fit first one data, then another and when you do that, it doesn't fit the first data anymore. Shouldn't just taking a single bigger sample provide better results? 
If you're into finding out, I can pull a commit that helps doing bigger samples via GPU. Already made it.

**Comment by ducha-aiki** [comment]
One iteration might be enough. Several iteration is safety
against difference in variance between batches (it happens sometimes, if
your data is not properly shuffled), but may be skipped the most of times.

**Comment by ibmua** [comment]
So, basically, it's about finding a sweetspot inside data where two consecutive batches would activate inputs so that they would generate same high-level pattern. Seems like using a 1 bigger batch should really be better than that. 
BTW, had been experimenting with normalization (also adjusting biases -= mean), neuron-level normalization and LSUV instead of layer-level, but had worse performance than your traditional LSUV for some reason.. Maybe, I should have adjusted learning rates, had LR of 0.001 for both tests, which was far from maximum, or maybe it really does lead to a worse performance.. Will explore that matter. Do you think neuron-level deviation adjustment should give better results than layerwise? And is it a good idea to adjust biases based on mean activation pattern of a neuron, or layer?

**Comment by ducha-aiki** [comment]
Probably you are right, I haven`t explored this. Just don`t trust single run methods :)
As for the per-neuron normalization - from the Orthonormal init, filters make some basis. LSUV just scale it. Doing this as per-neuron will change relative length of the vectors, this could be the issue. 
About minus mean. What activation are you using? If ReLU, I am afraid, that -= mean will lead to the more dead neurons. If you use leaky ReLU or ELU and it still worse, then I have no idea.

**Comment by ibmua** [comment]
Yes, I'm using PReLUs and ELUs. I got no idea why most people still use ReLUs in their nets. It's plain garbage. It just dies. While as far as I've tried PReLUs you can do same functions with them as can be done with Maxout and they learn them easier. Tried 
1 input neuron -> linear [1->6xOut] -> PReLU/ELU -> split in half (2x 3xOut) -> linear [3x->1xOut] for both -> Concatenate -> PReLU/ELU/Softmax -> [linear 2->1 output neuron] -> 1 output neuron.
This learned a non-convex function much better than Maxout -> Minout version did --- under much wider learning rate spectrum and, also, ELUs could make much more than just straight lines.

Meanwhile 
1 input neuron -> linear [1->6xOut] -> PReLU/ELU -> split in half (2x 3xOut) -> linear [3x->1xOut] for both -> Concatenate -> ReLU -> [linear 2->1 output neuron] -> 1 output neuron
would simply "die" and be a straight horizontal line, maybe, with a bias.

That's basically what Network In Network arch is built upon, I think. Although, they didn't look at it from same perspective as I did when I experimented with this after reading on Maxouts. I thought, what Maxout gives is a possibility of a new shape, and that can be also done with this sort of network. And it tuned out that, at least on smallest scale, it learns better, probably, because a line can't fall underneath two, or more other lines and not learn at all. Every neuron always learns a bit this way.

**Comment by ducha-aiki** [comment]
Could you please make this kind of activation for caffenet, so I can add your idea to the benchmark? :) https://github.com/ducha-aiki/caffenet-benchmark/blob/master/Activations.md

**Comment by ibmua** [comment]
Okay, I will, though I think that NIN arch is a more computationally-meaningful way that this is done. It's basically the same thing, except where it really doesn't give almost any additional computational complexity, it has a crossover between neurons which lead to, I think, for one thing more complexity which can degrade learning, though just a little bit, but for another it should lead to less need for repetition of same sort of neurons. Like if you have two features D and E and feature D depends on feature A and B while E depends on A and C, if you follow this plain "activation"-like Maxout approach, you will have to learn feature A two times, while if they are cross-connected like in NIN, you only need to learn to recognize A on a single neuron. I predict that while the wider Maxout-type network will perform a bit better and converge under higher LR without getting overfit, it will have to be much wider than if you connect those two consecutive kernel:1 layers to learn same complexity.

Okay, so I'll try to give you that Caffenet version soon. Will also send you an update to LSUV init, probably even sooner.

**Comment by ibmua** [comment]
Regarding interference of per-neuron init with Ortho init, I think what should probably matter most is how weights are initialized relative to one another. If we just scale all of them together, while they don't interfere with one another, as they are in the same layer, I think it should be okay. We just scale them so they all stay in some relative bounds and get overexcited, so we can set learning rates higher. You can also look at it from a perspective that any layer could be decomposed into 1 neuron length layers and put in parallel.
But I may be wrong. =) Testing on my data..

**Comment by ibmua** [comment]
I pulled for merge, but I've actually got my maths a bit wrong, I just understood that. Will fix tomorrow. It only matters for when one wants to adjust mean. It doesn't account for how mean changes due to weight scaling. One has to calculate how weights influenced mean.

Update: sent an update to that addresses the mean shift problem.

**Comment by ibmua** [comment]
Regarding -mean, it seems like both your paper and paper on Orthogonal init all completely omit caring about bias, so I think it's should be an interesting topic to explore.

**Comment by n3011** [comment]
@ducha-aiki hi its not working for PRELU layer. any solution??

**Comment by ducha-aiki** [comment]
@n3011 you should add to PReLU layer name "_act".
https://github.com/ducha-aiki/LSUVinit/blob/master/tools/extra/lsuv_init.py#L43

**Comment by n3011** [comment]
@ducha-aiki any better option than renaming?

**Comment by ducha-aiki** [comment]
@n3011 for now - not. One can check layer type in LSUV script, but problem is that loop is for blobs, not layers. Blob does not have layer type and blob name != layer name. So it is not so easy to fix.

**Comment by n3011** [comment]
@ducha-aiki okay thanks

**Comment by ibmua** [comment]
It's fixable via side parsing of prototxt, but yeah, unless someone else writes that one. =)
